src/THORSwap/Swap.cpp

```diff
@@ -0,0 +1,232 @@
+            return std::make_pair(std::move(out), std::move(res));
+        }
+
+        case Chain::BNB: {
+            Data out;
+            auto res = buildBinance(toChain, toSymbol, toTokenId, fromAddress, toAddress, vaultAddress, fromAmountNum, memo, out);
+            return std::make_pair(std::move(out), std::move(res));
+        }
+
+        case Chain::THOR:
+        default:
+            return std::make_pair<Data, std::string>({}, "Invalid from chain: " + std::to_string(toChain));
+    }
+}
+
+std::string Swap::buildBitcoin(Chain toChain, const std::string& toSymbol, const std::string& toTokenId, const std::string& fromAddress, const std::string& toAddress, const std::string& vaultAddress, uint64_t amount, const std::string& memo, Data& out) {
+    auto input = Bitcoin::Proto::SigningInput();
+
+    // Following fields need to be set after building, before sending
+    input.set_hash_type(TWBitcoinSigHashTypeAll);
+    input.set_byte_fee(1);
+    input.set_use_max_amount(false);
+    // private_key
+    // utxo
+    // scripts
+    // ... end
+
+    input.set_amount(amount);
+    input.set_to_address(vaultAddress);
+    input.set_change_address(fromAddress);
+    input.set_coin_type(TWCoinTypeBitcoin);
+    input.set_output_op_return(memo);
+
+    auto serialized = input.SerializeAsString();
+    out.insert(out.end(), serialized.begin(), serialized.end());
+    return "";
+}
+
+Data ethAddressStringToData(const std::string& asString) {
+    Data asData(20);
+    if (asString.empty() || !Ethereum::Address::isValid(asString)) {
+        return asData;
+    }
+    auto address = Ethereum::Address(asString);
+    std::copy(address.bytes.begin(), address.bytes.end(), asData.data());
+    return asData;
+}
+
+std::string Swap::buildEthereum(Chain toChain, const std::string& toSymbol, const std::string& toTokenId, const std::string& fromAddress, const std::string& toAddress, const std::string& vaultAddress, const std::string& routerAddress, uint64_t amount, const std::string& memo, Data& out) {
+    auto input = Ethereum::Proto::SigningInput();
+
+    // some sanity check / address conversion
+    Data vaultAddressBin = ethAddressStringToData(vaultAddress);
+    if (!Ethereum::Address::isValid(vaultAddress) || vaultAddressBin.size() != Ethereum::Address::size) {
+        return "Invalid vault address: " + vaultAddress;
+    }
+    if (!Ethereum::Address::isValid(routerAddress)) {
+        return "Invalid router address: " + routerAddress;
+    }
+    Data toAssetAddressBin = ethAddressStringToData(toTokenId);
+
+    // Following fields need to be set after building, before sending
+    auto chainId = store(uint256_t(1));
+    input.set_chain_id(chainId.data(), chainId.size());
+    auto nonce = store(uint256_t(11));
+    input.set_nonce(nonce.data(), nonce.size());
+    auto gasPrice = store(uint256_t(10000000000));
```

Comment:
why all the values are hardcoded?

Reply:
Good catch.  These fields must be set later, so these values are just placeholders here.  Changed them to 0 to be more clear.